### PR TITLE
Pv2 2268 redundancy bil funding source bug

### PR DESCRIPTION
**Pull request by tom-gough**

### CODE REVIEW

File: src/SFA.DAS.Payments.EarningEvents.Application/Mapping/EarningEventBuilderBase.cs [hunk 1]
```diff
@@ -96,5 +97,37 @@ private static bool IsCurrentAcademicYear(PriceEpisodeValues priceEpisodeValues,
...
+
+            var orderedPriceEpisodes = priceEpisodes.OrderBy(x => x.PriceEpisodeValues.EpisodeStartDate).ToList();
+            var priceEpisodeCount = orderedPriceEpisodes.Count;
+
+            for (var i = 0; i < priceEpisodeCount; i++)
+            {
+                var currentPriceEpisode = orderedPriceEpisodes[i];
+                if (currentPriceEpisode.PriceEpisodeValues.PriceEpisodeRedStatusCode == 1)
+                {
+                    var startPeriod = currentPriceEpisode.PriceEpisodeValues.PriceEpisodeRedStartDate.Value.GetPeriodFromDate();
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.EarningEvents.Application/Mapping/EarningEventBuilderBase.cs:116]: I would rename startPeriod to redundancystartPeriod and nextPeriod to redundancyStoppedPeriod or redundancyEndPeriod or redundancyFinishedPeriod ?

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs [hunk 2]
```diff
@@ -18,56 +17,53 @@ public RedundancyEarningService(IRedundancyEarningEventFactory redundancyEarning
...
         {
-            var splitResults = new List<ApprenticeshipContractTypeEarningsEvent>();
-          
-            var redundancyPeriod = redundancyDate.GetPeriodFromDate(); 
+            var splitResults = new List<ApprenticeshipContractTypeEarningsEvent> {earningEvent};
+            if (!redundancyPeriods.Any())
+                return splitResults;
 
             var redundancyEarningEvent = redundancyEarningEventFactory.CreateRedundancyContractTypeEarningsEvent(earningEvent);
+            splitResults.Add(redundancyEarningEvent);
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:28]: even though the end result is the same I would move this line at the bottom to make it clear that 
split results contains updated redundancy earnings, here its bit lost in other code and someone might not see it

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs [hunk 3]
```diff
@@ -18,56 +17,53 @@ public RedundancyEarningService(IRedundancyEarningEventFactory redundancyEarning
             this.redundancyEarningEventFactory = redundancyEarningEventFactory ?? throw new ArgumentNullException(nameof(redundancyEarningEventFactory));
         }
 
-        public List<ApprenticeshipContractTypeEarningsEvent> SplitContractEarningByRedundancyDate(ApprenticeshipContractTypeEarningsEvent earningEvent, DateTime redundancyDate)
+        public List<ApprenticeshipContractTypeEarningsEvent> SplitContractEarningByRedundancyDate(ApprenticeshipContractTypeEarningsEvent earningEvent,
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:20]: I would rename the method to reflect the new behaviour, 1 its not doing the split by date and 2 its not doing split at all  in one instance (reinstating that if statement would mean method will always be doing split)

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs [hunk 4]
```diff
@@ -18,56 +17,53 @@ public RedundancyEarningService(IRedundancyEarningEventFactory redundancyEarning
...
 
-            splitResults.Add(earningEvent);
-            splitResults.Add(redundancyEarningEvent);
-            
             return splitResults;
         }
 
-        public List<FunctionalSkillEarningsEvent> SplitFunctionSkillEarningByRedundancyDate(FunctionalSkillEarningsEvent functionalSkillEarning,
-            DateTime priceEpisodeRedStartDate)
+        public List<FunctionalSkillEarningsEvent> SplitFunctionSkillEarningByRedundancyDate(FunctionalSkillEarningsEvent functionalSkillEarning, List<byte> redundancyPeriods)
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:47]: same as SplitContractEarningByRedundancyDate

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs [hunk 5]
```diff
@@ -18,56 +17,53 @@ public RedundancyEarningService(IRedundancyEarningEventFactory redundancyEarning
...
         {
-            var splitResults = new List<FunctionalSkillEarningsEvent>();
-            var redundancyPeriod = priceEpisodeRedStartDate.GetPeriodFromDate();
+            var splitResults = new List<FunctionalSkillEarningsEvent>{ functionalSkillEarning };
+            if (!redundancyPeriods.Any())
+                return splitResults;
 
             var redundancyEarningEvent = redundancyEarningEventFactory.CreateRedundancyFunctionalSkillTypeEarningsEvent(functionalSkillEarning);
-            
+            splitResults.Add(redundancyEarningEvent);
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:54]: same as SplitContractEarningByRedundancyDate

**Review by hardik-desai-kcl — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/PV2_2268_Steps.cs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+            await WaitForPayments(2);
+        }
+
+        [Given("a break in learning has also occurred")]
+        [Given("the redundancy and break in learning have been correctly recorded in the ILR")]
+        [Given("the delivered learning days before the break are under 75 percent")]
+        public void EmptyStep()
+        {
+            //covered by FM36s imported in first step
```
[the remark this answers is not in the document]
Comment (tom-gough) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/PV2_2268_Steps.cs:96]: done

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/PV2_2268_Steps.cs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                .Any(x =>
+                    x.FundingSourceType == FundingSourceType.Levy
+                    && x.DeliveryPeriod == 4
+                    && x.TransactionType == TransactionType.Learning);
+        }
+
+        private void ImportR03Fm36() { GetFm36LearnerForCollectionPeriod("R03/current academic year"); }
+        private void ImportR04Fm36() { GetFm36LearnerForCollectionPeriod("R04/current academic year"); }
+
+        private void ScriptInApprenticeshipData()
```
[the remark this answers is not in the document]
Comment (tom-gough) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/PV2_2268_Steps.cs:119]: done

File: src/SFA.DAS.Payments.EarningEvents.Application/Mapping/EarningEventBuilderBase.cs (hunk 1 above)
Comment (tom-gough) [reply src/SFA.DAS.Payments.EarningEvents.Application/Mapping/EarningEventBuilderBase.cs:116]: done

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs (hunk 2 above)
Comment (tom-gough) [reply src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:28]: That doesn't really match the approach we have gone for with the ApprenticeshipContractTypeEarningsEvent. The idea is we add both earning events (if required) to the list at the top, then set them up correctly in the rest of the method. If we were to move this to the bottom we would be adding the earning event at the top, but then the redundancy earning event at the bottom. If we were to move both to the bottom we would have to change the clean "return early" code on lines 23-25

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs (hunk 3 above)
Comment (tom-gough) [reply src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:20]: done

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs (hunk 4 above)
Comment (tom-gough) [reply src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:47]: done

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs (hunk 5 above)
Comment (tom-gough) [reply src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:54]: done

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/ApprenticeshipContractTypeEarningsEventBuilderTest.cs [hunk 8]
```diff
@@ -736,6 +740,386 @@ public void CreatedTheCorrectNumberOfEarningEventsWhenLearningSupportIncluded()
...
+                .OnProgrammeEarnings
+                .Single(x => x.Type == OnProgrammeEarningType.Learning)
+                .Periods
+                .Should().HaveCount(1);
+            events.Where(x => x.PriceEpisodes.Exists(y => y.Identifier == "25-17-01/08/2020"))
+                .Single(x => x.GetType().IsAssignableFrom(typeof(ApprenticeshipContractType1RedundancyEarningEvent)))
+                .IncentiveEarnings
+                .Single(x => x.Type == IncentiveEarningType.LearningSupport)
+                .Periods
+                .Should().HaveCount(1);
```
[the remark this answers is not in the document]
Comment (tom-gough) [reply src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/ApprenticeshipContractTypeEarningsEventBuilderTest.cs:1126]: sensible ye, done

**Review by hardik-desai-kcl — APPROVED** [review]

**Pull request merged**